### PR TITLE
Change rustdoc-scrape-examples to be a target-level configuration

### PR DESCRIPTION
**Pull request by willcrichton**
This PR addresses issues raised in rust-lang/cargo#9525. Specifically:
1. It enables examples to be scraped from `#[test]` functions, by passing additional flags to Rustdoc to ensure that these functions aren't ignored by rustc. 
2. It moves the `arg` from `-Z rustdoc-scrape-examples={arg}` into a target-level configuration that can be added to Cargo.toml. 

The added test `scrape_examples_configure_target` shows a concrete example. In short, examples will be default scraped from Example and Lib targets. Then the user can enable or disable scraping like so:

```toml
[lib]
doc-scrape-examples = false

[[test]]
name = "my_test"
doc-scrape-examples = true
```

### CODE REVIEW

**Comment by camelid** [comment]
Somehow I feel like it'd be better to structure the configuration as something like:

```
[doc]
scrape-examples-targets = ["lib", "test"]
```

rather than putting `doc-scrape-example` in each section. Just a gut reaction though.

**Comment by willcrichton** [comment]
> Somehow I feel like it'd be better to structure the configuration as something like:
> 
> ```
> [doc]
> scrape-examples-targets = ["lib", "test"]
> ```
> 
> rather than putting `doc-scrape-examples` in each target section. Just a gut reaction though.

I think the issue with this is (a) it's not consistent with existing ways to configure targets, and (b) it doesn't allow target configuration at a higher granularity. For instance, the proposed method allows you to individually enable/disable specific examples and tests, while a list of target strings can't do so.

**Comment by alexcrichton** [comment]
This all looks pretty reasonable to me, but I'd like to confirm my understanding. The intention is that once stabilized this is enabled for all projects by default? So all projects with libraries and examples, when using `cargo doc`, will see examples for methods and such drawn from libraries examples. Packages have fine-grained control at the per-target level what's used as candidates for examples and what isn't (with the default being only the lib targets and examples)

If that's the case I think it might be good to flesh out the failing tests for now. I think there may need to be some trickery around not running doc-scrape with stable Rust as-is today since Cargo tests its CI with all three channels of Rust. Otherwise though that seems pretty reasonable to me and the PR otherwise looks pretty good.

>What is the right set of default targets to search for examples?

Personally the choice made here I think is fine. This seems like something we can alter over time if necessary as well.

> Is there a different granularity at which example scraping should also be configurable?
 
At least from Cargo's perspective this probably won't have much impact, but I don't have much of an opinion on this otherwise. 


----

As a random bikeshedding point seeing `doc-scrape-examples = false` in isolation feels a little surprising, so I might suggest something like `doc-example-candidate = false` or something like that.

**Comment by GuillaumeGomez** [comment]
I thought this feature would be disabled by default?

**Comment by willcrichton** [comment]
> I thought this feature would be disabled by default?

My understanding is that it would be disabled by default while unstable, and then enabled by default once stabilized. The vast majority of people will not use this feature if it is hidden behind a flag.

**Comment by GuillaumeGomez** [comment]
> My understanding is that it would be disabled by default while unstable, and then enabled by default once stabilized. The vast majority of people will not use this feature if it is hidden behind a flag.

Sorry, I definitely didn't understand it like this. In this case, we'll need to update the display quite a lot because it takes a lot of space by default. What about collapsing the section by default with a title like "see usage in examples" or something among the line? If you prefer we can discuss it on zulip.

**Comment by willcrichton** [comment]
The discussion about scraping-by-default-or-not has been pulled onto Zulip: https://rust-lang.zulipchat.com/#narrow/stream/266220-rustdoc/topic/Enabling.20scrape.20examples.20by.20default

**Comment by lqd** [comment]
@willcrichton The `.rustfmt.toml` change and new `log` file, seem unintentional.

**Comment by willcrichton** [comment]
Ah thanks @lqd , got caught up in the rebase. I'll squash everything in a bit.

**Comment by willcrichton** [comment]
This PR is ready for review, although it needs a new Cargo reviewer since Alex has stepped back.

As a reminder, this PR will enable the scrape examples feature by default, although only when run with a nightly rustc. Now that rust-lang/rust#93217 has landed, I think this feature is ready to be rolled out at least on docs.rs. Rustdoc maintainers, please let me know if you have any remaining concerns about this feature. 

cc @jsha @camelid @GuillaumeGomez

@rustbot ready

**Comment by willcrichton** [comment]
It seems like only @ehuss is on [reviewer rotation](https://github.com/rust-lang/highfive/blob/master/highfive/configs/rust-lang/cargo.json) for Cargo right now, so I will add him. Sorry for all the extra work Eric :(

r? @ehuss

File: src/cargo/ops/cargo_compile.rs
Comment (ehuss) [inline src/cargo/ops/cargo_compile.rs:376]: There are a lot of local variables in this function, and I'm having a hard time keeping them straight. (spec, specs, full_specs, resolve_specs, need_reverse_dependencies).

Might I suggest a few things to maybe clarify it a little? Maybe something like:

```rust
let needs_scrape_examples =
    matches!(build_config.mode, CompileMode::Doc { .. }) && config.nightly_features_allowed;
let requested_specs = spec.to_package_id_specs(ws)?;
let specs_to_resolve = if needs_scrape_examples {
    // When scraping examples, this neads All because …
    Packages::All.to_package_id_specs(ws)?
} else {
    requested_specs.clone()
};
```

Where this:
* `need_reverse_dependencies` → `needs_scrape_examples`
* `resolve_specs` → `specs_to_resolve`
* Remove `full_specs` and `specs`
* Add `requested_specs` for the specs requested by the user

Also, I'm a bit surprised that scraping forces scraping from the entire workspace. That seems like it could cause a lot of problems. Am I understanding what this is doing correctly?

**Comment by ehuss** [comment]
This seems to be a very unusual way of stabilizing something. Can you clarify exactly how this is intended to work? I'm skeptical that our setup can handle stabilizing in this fashion, so I'm a little concerned. Is there an intent to stabilize the rustdoc side? Usually that is done before the cargo side.

I'm also a little surprised to see this moving for stabilization so quickly. I don't have much of an impression that it has gotten much testing. I also haven't seen much discussion from the rustdoc side.

File: src/cargo/core/manifest.rs
Comment (willcrichton) [reply src/cargo/core/manifest.rs]: Sure, I implemented your suggested strategy in the most recent commit.

File: src/cargo/core/compiler/job_queue.rs
Comment (willcrichton) [reply src/cargo/core/compiler/job_queue.rs]: I implemented the strategy you suggested. But if you don't think it's worth it to notify the user about scraping, I can also remove the notification.

File: src/cargo/core/compiler/unit.rs
Comment (willcrichton) [reply src/cargo/core/compiler/unit.rs]: This was to ensure that the buildkey was stable across sessions, but this fact is no longer needed.

File: src/cargo/core/features.rs
Comment (willcrichton) [reply src/cargo/core/features.rs:890]: Good idea, I added that in the latest commit.

File: src/cargo/ops/cargo_compile.rs
Comment (willcrichton) [reply src/cargo/ops/cargo_compile.rs]: I think the recent bugfix PRs addressed this issue, so I reverted this change.

File: src/cargo/ops/cargo_compile.rs
Comment (willcrichton) [reply src/cargo/ops/cargo_compile.rs]: Ok yes, I think metadata is actually what I needed. So I changed this feature to use `metadata` instead of `buildkey`, and I also added a test to ensure that scraping is properly memoized. I reverted all changed to unit hashing, since they're no longer necessary.

File: src/cargo/ops/cargo_compile.rs
Comment (willcrichton) [reply src/cargo/ops/cargo_compile.rs]: I tried to clean up the code roughly in line with your suggestion. Let me know if you think it's clearer.

And yes, scraping does by default look through all the packages in your workspace for examples. What kind of problems are you anticipating?

**Comment by willcrichton** [comment]
My goal here is to provide a bigger roll out for this feature before fully stabilizing it. There's been a lot of back-and-forth on the UI, but those discussions were finally settled (for now at least) in rust-lang/rust#93217. Essentially this change will activate the scrape examples feature by default on docs.rs, which uses nightly. More users will encounter the UI, and we we will be able to identify any remaining Cargo issues from broken doc builds. Once users are satisfied with the UI and all Cargo issues are settled, then we can finish stabilizing the feature.

Sorry, it may be misleading to say this PR "prepares the feature for stabilization". I mean preparing in a looser sense, as in "is the biggest remaining step before stabilization"

File: src/cargo/core/compiler/job_queue.rs
Comment (ehuss) [reply src/cargo/core/compiler/job_queue.rs]: This still doesn't seem to be working as I would expect.

In a project with `src/lib.rs` and `examples/foo.rs`, `cargo doc` doesn't mention "Scraping". I think the filter:

```rust
self.compiled.contains(&unit.pkg.package_id()) && !unit.mode.is_doc()
```

would need to be:

```rust
self.compiled.contains(&unit.pkg.package_id()) && !unit.mode.is_doc() && !unit.mode.is_doc_scrape()
```

As for whether or not to do it at all, I'm on the fence. We have felt that cargo is already a little too verbose (eventually we will likely get rid of all the status messages during a build).

If it were me, I would probably not include it.

File: src/cargo/ops/cargo_compile.rs
Comment (ehuss) [reply src/cargo/ops/cargo_compile.rs]: I'm pretty sure that isn't something we can do by default. Workspaces will have packages that don't compile, or only build on certain targets, or have very expensive build scripts, or any number of requirements. Some workspaces may be huge, but the user only wants to document one crate. There are some users who have monolithic workspaces across entire orgs.

I think if the user wants to document the entire workspace, they can specify `--workspace` or multiple `-p` flags.

**Comment by ehuss** [comment]
This does seem to actually stabilize it, though, in the sense that there is no longer a `-Z` flag to gate it. Cargo specifically has a system to make things like this require an explicit opt-in before it is stabilized. Users should be able to use the nightly channel the same as the stable channel.

I might suggest a different way to try to get more exposure. Would it be possible to just enable it on docs.rs by default? Have you talked to them about it?

Also, another tactic is to do a crater run with it enabled, just to see what basic problems might arise.  I would do that before docs.rs.  (I would also make sure the issues raised in this PR are fixed first.)

File: src/cargo/core/compiler/job_queue.rs
Comment (ehuss) [reply src/cargo/core/compiler/job_queue.rs]: I just realized that it does need to print *something*, otherwise it will sometimes say "Compiling" which probably isn't good.  I think with the above adjustment, it should probably be fine.

Comment (willcrichton) [reply src/cargo/core/compiler/job_queue.rs]: Ah ok I see. I've fixed it with what you suggested, and it consistently indicates when scraping.

File: src/cargo/ops/cargo_compile.rs
Comment (willcrichton) [reply src/cargo/ops/cargo_compile.rs]: Ok, that makes sense. I changed it to just use the package set requested by the user, and we can eventually document that users should do `--all` or whatever if they want to scrape all possible examples.

**Comment by willcrichton** [comment]
Ok that makes sense. Since this isn't stable yet, I restored the `-Zrustdoc-scrape-examples` flag. But now instead of passing `=all` or `=examples`, that configuration is still done at the target level in Cargo.toml. I will also do a crater run this weekend to ferret out any issues. I'll check with the docs.rs folks and see if they're willing to add the flag.

**Comment by willcrichton** [comment]
Update: I did a crater run of the scrape-examples feature as it exists on nightly on the top-10,000 crates (after implementing rust-lang/crater#626). In total there were 131 regressions. I could not find any issues directly in the implementation of scrape-examples. Instead, regressions had one of three causes:

1. The project's examples did not type-check.
2. The project itself did not type-check (e.g. `cargo check` would fail while `cargo doc` would work).
3. Dependencies of the examples did not type-check.

Aside: I'm actually a bit surprised that `cargo publish` includes examples by default, but doesn't check if they compile. Is that intended behavior? Ensuring that examples work at publish-time would avoid the first issue.

So the question then is whether these regressions matter. If it's not important that `cargo doc` should still work in these cases, then I think the feature is ready for deployment on docs.rs, especially once this PR lands. But if it _is_ important, then I think the only fix would be to somehow mark the scrape units in Cargo as "fallible" or "optional", like if they fail, the build should still continue. I'm not sure if Cargo has such a concept, or how easy it would be to implement. 

@ehuss what are your thoughts on this?

**Comment by willcrichton** [comment]
After some discussion with @jyn514 on Zulip, it's clear that a `Docscrape` unit should be able to fail without killing the build. Some crates (but not most!) will rely on the fact that `cargo doc` will work while `cargo check` will not, and that is fundamental to the design of rustdoc.

So I think what I'm going to do is keep this PR scoped to its original goal, of changing the target configuration for the scrape-examples feature (plus the couple bugfixes that have emerged in review). Then I'm going to put up a second PR that adds the ability to mark units as "ok to fail", which will be like a more granular version of `BuildConfig::keep_going`, and mark all Docscrape units as such. So @ehuss let me know if you have any more comments on this PR as is.

**Comment by willcrichton** [comment]
@ehuss I just implemented the strategy you described [here](https://github.com/rust-lang/cargo/pull/10596#issuecomment-1186773249), and added a set of corresponding unit tests. Let me know if you think this captures your intent.

One implementation note: I'm using `state.emit_diag("warning", ...)` to emit the warning when silenced scraped units fail. I'm not sure if that's the correct API to use.

**Comment by ehuss** [comment]
Can we move all of the docscrape tests to a separate module?  Having `doc.rs` climb past 3,000 lines is making it a bit too heavy-weight, and it can be convenient to keep new features in their own little scope. I'll leave it up to you if it would be easier to do before or after this PR.

**Comment by ehuss** [comment]
There doesn't seem to be anything that will enabling scraping of targets with `doc-scrape-examples=true` set (or to disable it if `doc-scrape-examples=false`). 

I see that may be difficult to implement because `has_dev_units` needs to be determined before running the resolver, but it doesn't know what is necessarily being built before running the resolver. Perhaps one workaround is to call `ws.members_with_features(…)` and check if any of those packages have example targets with `doc-scrape-examples=true`, and if so enable dev-dependencies, and change the `CompileFilter` to include those targets.

Conversely, I imagine there should be some filter somewhere to prevent targets with `doc-scrape-examples=false` from being implicitly included.

File: src/cargo/ops/cargo_compile.rs
Comment (willcrichton) [reply src/cargo/ops/cargo_compile.rs]: Ah yeah that's right. Using `to_builds` seems reasonable, I changed this function to use that set instead of `PackageSet`.

File: src/cargo/core/compiler/mod.rs
Comment (willcrichton) [reply src/cargo/core/compiler/mod.rs:652]: It isn't possible in that you can't pass `--tests` to `cargo doc`. But re: later in your review, if we make it so `[[tests]] doc-scrape-examples = true` will proactively include the test target, then this could arise. So it's a bit of future proofing.

File: src/cargo/core/compiler/mod.rs
Comment (willcrichton) [reply src/cargo/core/compiler/mod.rs]: Good point, I think just setting `show_diagnostics` is the simpler approach. That at least passes my simple test.

File: src/cargo/core/compiler/mod.rs
Comment (willcrichton) [reply src/cargo/core/compiler/mod.rs]: Done.

File: src/cargo/core/compiler/mod.rs
Comment (willcrichton) [reply src/cargo/core/compiler/mod.rs:763]: Rustdoc allows functions without valid bodies. If you change it to something like `fn main();` then Rustdoc will return a nonzero exit code.

File: src/cargo/core/compiler/mod.rs
Comment (willcrichton) [reply src/cargo/core/compiler/mod.rs]: Ok, I implemented the strategy you suggested.

One issue though is I'm not sure how this special pathway should interact with the deduplication functionality. It seems like deduplication is handled both within the child process and the parent process?

**Comment by willcrichton** [comment]
> Can we move all of the docscrape tests to a separate module? Having doc.rs climb past 3,000 lines is making it a bit too heavy-weight, and it can be convenient to keep new features in their own little scope. I'll leave it up to you if it would be easier to do before or after this PR.

Done, moved to a new `docscrape.rs`.

> There doesn't seem to be anything that will enabling scraping of targets with doc-scrape-examples=true set (or to disable it if doc-scrape-examples=false).
> I see that may be difficult to implement because has_dev_units needs to be determined before running the resolver, but it doesn't know what is necessarily being built before running the resolver. Perhaps one workaround is to call ws.members_with_features(…) and check if any of those packages have example targets with doc-scrape-examples=true, and if so enable dev-dependencies, and change the CompileFilter to include those targets.

I just added that here, based on the strategy you recommended:

https://github.com/rust-lang/cargo/pull/10343/files#diff-98da1b66b532e50e9bca971b453ee7de96b17436de1d7ef3824c9888d55be9bbR377-R395

> Conversely, I imagine there should be some filter somewhere to prevent targets with doc-scrape-examples=false from being implicitly included.

That is the purpose of this check after running `generate_targets`:

https://github.com/rust-lang/cargo/pull/10343/files#diff-98da1b66b532e50e9bca971b453ee7de96b17436de1d7ef3824c9888d55be9bbR535-R543

**Comment by willcrichton** [comment]
I did another Crater run on the top 10k packages with the latest scrape examples implementation. I realized that the current strategy does not handle the following case:

1. A crate's library *documents* but does not *compile*.
2. The crate has an example and no dev-dependencies.
3. Cargo tries to opportunistically scan the example.
4. To scan the example, Cargo needs to check the example's dependencies, which includes the crate library.
5. Cargo compiles the crate library and fails. The lib-check unit is *not* opportunistic, so its failure crashes the entire build.

I can only think of a few solutions to this:
1.  **General solution:** essentially what I proposed in #10596. If *any transitive dependency* of a fallible (or opportunistic or what have you) unit fails, then that has to kill the build *up to the fallible dependency*. This will solve the problem robustly, but adds a lot of new concepts and implementation complexity.
2. **Specific solution:** if a lib-check unit fails AND it's a dependency of a scrape unit, then don't crash the build. This solves the specific problem, but adds yet another edge case to Cargo.
3. **Scrape examples becomes opt-in:** we give up on trying to make any kind of default behavior for scrape examples, and users have to opt-in by passing some kind of flag or Cargo.toml parameter. This is guaranteed backward-compatible, but would cause the feature to be used far less.

@ehuss What do you think strikes the right balance between feature impact and Cargo tech debt? Or do you see another possible solution?

**Comment by jyn514** [comment]
@willcrichton just fyi - I would recommend trying to land the changes you have so far before trying to make more. This PR is already quite big and it's hard to skim the comments, having a separate issue for the problem you found would be helpful I think.

**Comment by willcrichton** [comment]
That's sensible, I don't want to expand the scope of this PR too much. I'm happy to have this PR be merged as-is if we want to pull this discussion elsewhere.

**Comment by willcrichton** [comment]
@ehuss this PR is ready for review whenever you have the time. I'll fix the issue described in https://github.com/rust-lang/cargo/pull/10343#issuecomment-1203170975 in a later PR.

**Comment by willcrichton** [comment]
@ehuss do you still have the cycles to review this PR? If not that's totally fine, I can ask epage or weihanglo to take over.

File: src/cargo/core/compiler/job_queue.rs
Comment (willcrichton) [reply src/cargo/core/compiler/job_queue.rs]: I have added an explanation, let met know if it makes sense.

File: src/cargo/core/compiler/rustdoc.rs
Comment (willcrichton) [reply src/cargo/core/compiler/rustdoc.rs:197]: I have added a doc comment, let me know if it makes sense.

File: src/cargo/core/manifest.rs
Comment (willcrichton) [reply src/cargo/core/manifest.rs]: Done.

File: src/cargo/core/compiler/context/mod.rs
Comment (willcrichton) [reply src/cargo/core/compiler/context/mod.rs:84]: I have added a doc comment, let me know if it makes sense.

File: src/cargo/ops/cargo_compile/mod.rs
Comment (willcrichton) [reply src/cargo/ops/cargo_compile/mod.rs]: Done.

File: tests/testsuite/package_features.rs
Comment (willcrichton) [reply tests/testsuite/package_features.rs:471]: This change causes the error to happen earlier. https://github.com/rust-lang/cargo/pull/10343/files#diff-9b38573ab038c139588eedf8c9e475e5ff047deb3b8335ca4fa94642e9b31fe6R238-R267

File: src/cargo/core/compiler/job_queue.rs
Comment (willcrichton) [reply src/cargo/core/compiler/job_queue.rs:444]: I added a reference back to the explanation on the relevant enum.

File: tests/testsuite/docscrape.rs
Comment (willcrichton) [reply tests/testsuite/docscrape.rs]: This is from #10500, I changed the name to reflect that.

File: tests/testsuite/docscrape.rs
Comment (willcrichton) [reply tests/testsuite/docscrape.rs:408]: I added an explanation, let me know if it makes sense.

File: src/cargo/ops/cargo_compile/mod.rs
Comment (willcrichton) [reply src/cargo/ops/cargo_compile/mod.rs]: I added a lengthy explanation (that also relates to the `no_scrape_with_dev_deps` test). Let me know if it makes sense.

**Comment by willcrichton** [comment]
@weihanglo thanks for the review. I've either fixed or responded to all of your comments.

**Comment by weihanglo** [comment]
This would fix https://github.com/rust-lang/cargo/issues/10876 as well, right?

**Comment by willcrichton** [comment]
> This would fix #10876 as well, right?

That's correct. I just added the suggested change to the `docscrape::basic` test to confirm this.

File: src/cargo/ops/cargo_compile/mod.rs
Comment (weihanglo) [reply src/cargo/ops/cargo_compile/mod.rs]: I love your writing. Thank you for providing such a good explanation!

**Review by weihanglo — APPROVED** [review]
Thanks for the update! I believe this pull request is effectively an implementation based on ehuss's comment <https://github.com/rust-lang/cargo/pull/10596#issuecomment-1186773249>, plus fingerprint support for scrape units. Totally agree it is the best effort to roll out this feature by default with minimal intrusion. 

Since this is still unstable, I am going to merge it without an FCP. If anyone thinks Cargo shouldn't scrape examples under those conditions, we can revisit it and change the logic of compile filters.

**Comment by willcrichton** [comment]
Seems like the test `lto::test_profile` failed, but I don't think anything in this PR would have affected it. Any reason to believe it's flaky?

**Comment by weihanglo** [comment]
I am not sure if it is related to https://github.com/rust-lang/rust/issues/98302, but [it can be reproduced](https://github.com/weihanglo/cargo/actions/runs/3541971990/jobs/5946901773) so seems not a flaky test?

Anyway, let's try again.

@bors retry

**Pull request merged**
